src/freenet/node/MultiMessageCallback.java

```diff
@@ -2,14 +2,37 @@
 
 import freenet.io.comm.AsyncMessageCallback;
 
-/** Waits for multiple asynchronous message sends, then calls finish(). */
+/** Waits for multiple asynchronous message sends, then calls finish(). 
+ * You should add messages with make() and then call arm(). sent(boolean)
+ * will be called when all the messages have been sent (or failed e.g. 
+ * disconnected) and finish(boolean) will be called when all the messages 
+ * have been acknowledged (or failed). Typical usage:
+ * <pre>Message m1 = ...;
+ * Message m2 = ...;
+ * PeerNode pn = ...;
+ * MultiMessageCallback mcb = new MultiMessageCallback() {
+ *   protected void finish(boolean success) {
+ *     // Messages have finished.
+ *   }
+ *   protected void sent(boolean success) {
+ *     // Messages have been sent.
+ *   }
+ * }
+ * pn.sendAsync(m1, mcb, ctr);
+ * pn.sendAsync(m2, mcb, ctr);
+ * mcb.arm();</pre> */
 public abstract class MultiMessageCallback {
 	
+    /** Number of messages that have not yet completed */
 	private int waiting;
+	/** Number of messages that have not yet been sent */
 	private int waitingForSend;
 	
+	/** True if arm() has been called. finish(boolean) and sent(boolean) will 
+	 * only be called after arming the callback. */
 	private boolean armed;
```

Comment:
Matthew Toseland writes:

>>  	private int waitingForSend;
>>  	
>> +	/** True if arm() has been called. finish(boolean) and sent(boolean) will 
>> +	 * only be called after arming the callback. */
>>  	private boolean armed;
>
> See below.
> 	/** Enable the callback. The callbacks sent(boolean) and finish(boolean)
> 	 * will only be called after this method has been called, so you should 
> 	 * use this to indicate that you won't add any more messages. */ 
> Is this unclear?

I missed that — I think it’s pretty clear. So arming means something
like finalizing a packet?

Best wishes,
Arne

-- 
Unpolitisch sein
heißt politisch sein
ohne es zu merken